active/003-seq-feature.md

```diff
@@ -0,0 +1,180 @@
+[70, 100)                                    {'name': 'cds1'}
+[50, 200)   {'db_xref': ('EcoGene:EG11555', 'GeneID:944757...
+[120, 150)     {'db_xref': 'EcoGene:EG11555', 'name': 'gen3'}
+[120, 150)                                   {'name': 'gen4'}
+   ```
+
+## Types of feature queries
+* query by location. `IntervalIndex` module implements `IntervalTree` for fast query by location. To get features covering a unit or a region, we can do this:
+
+``` python
+>>> interval_md.loc[30, ]
+          TYPE STRAND  LEFT RIGHT  \
+[0, 100)  gene      +  True  True
+
+                                                  ATTIBUTES
+[0, 100)  {'name': 'gen1', 'gene_synonym': 'ECK0007; JW0...
+>>> interval_md.loc[Interval(10, 30), ]
+          TYPE STRAND  LEFT RIGHT  \
+[0, 100)  gene      +  True  True
+[0, 30)    CDS      +  True  True
+
+                                                 ATTRIBUTES
+[0, 100)  {'name': 'gen1', 'gene_synonym': 'ECK0007; JW0...
+[0, 30)                                    {'name': 'cds1'}
+```
+
+* query by attributes. It can be used to fetch multiple features or a single feature. For example:
+
+``` python
+>>> interval_md[interval_md.apply(lambda row: row['ATTRIBUTES']['name']=='cds1', axis=1)]
+          TYPE STRAND  LEFT RIGHT        ATTRIBUTES
+[0, 30)    CDS      +  True  True  {'name': 'cds1'}
+[70, 100)  CDS      +  True  True  {'name': 'cds1'}
+
+```
+
+## Operations that need update in `interval_metadata`
+
+There are some methods in `Sequence` and its child classes that needs to operate on `interval_metadata` accordingly. For all the situations below, in case users only want to rc the sequence and do not intend to use `interval_metadata`, we should provide an option to skip changes and only return the sequence without `interval_metadata`.
+
+* Reverse operation. `INTERVAL` needs to be updated:
+``` python
+# original interval for a 9 nt gene
+>>> a = Interval(2, 11, closed='left')
+# after reverse complemenary of a seq of 12 nt, the interval should be
+>>> a_rc = 12 - a
+>>> a_rc
+Interval(1, 10, closed='left')
+```
+
+* Complementary operation. The `STRAND` needs to be switched.
+
+* Concatenation. The intervals of all the sequences, except for the 1st one, needs to be updated.
+
+``` python
+# interval of a feature in seq1 of 10 nt
+>>> a1 = Interval(2, 9, closed='left')
+# interval of a feature in seq2
+>>> a2 = Interval(1, 3, closed='left')
+# after concat, the two features will have the intervals of:
+>>> a1  # still the same
+Interval(2, 9, closed='left')
+>>> a2 + 10
+Interval(11, 13, closed='left')
+```
+
+* Slicing. The intervals needs to reflect the change of the sequence. Only keep the features that overlaps with sliced sequence, and need to update the coordiate as well. One caveat is that if a feature is cut in the middle of the slicing with missing left or right ends, the `LEFT` or `RIGHT` needs to be changed to `TRUE`
+
+# Alternatives
+
+Stumped here. Any ideas?
```

Comment:
We didn’t solve the general case (because the dev effort would have been very large and the actual use cases were few). We supported a lot of things like canOverlap, mustOverlap, canNotOverlap, mustNotOverlap, etc. and supported the critical feature of being able to extract the coding sequence from the DNA.

Note that there are a lot of biological phenomena like trans-splicing that make this really hard (trans splicing is when a feature spans multiple transcripts which can come from different strands of the DNA — yes, you can’t assume that everything in the transcript comes from the same strand — or even from different records; we didn’t solve that latter case because figuring out where to get the external record from (search memory? a file? a local database? NCBI?) was hard.

A good way to focus this discussion is to figure out what are the use cases we must support, what would be nice to support, and what we won’t support, and make pragmatic choices from there that allow us to read the “easy” GenBank files soon. Let’s face it, this will be 99.99% of our use cases, and the “hard” genbank files may never be supported unless someone has a specific research question that they’re required to solve (like in trans-splicing).

Note that even questions such as “is position x in an intron” are actually hard because this depends on what the possible splice junctions are, which have been experimentally observed and under what conditions, and do you mean “is it within bounds of an initial transcript but not always in the final transcript” or “is it within bounds of an initial transcript but not ever in the final transcript” etc. So part of what makes the decisions hard is that you run into unresolved biological definitions and/or questions rapidly. This is especially problematic with exon skipping and alternative start sites, as we found when trying to map DNA to proteome, because if you do it at the whole CDS level including hypothetical combinations of splice junctions you get a combinatorial explosion. Eventually we gave up and did a peptide-centric database instead…

So, my vote would be to do something that works for the easy cases (e.g. where everything is in the same file), handles the clear-cut parts of the syntax (e.g. where you know the start and end points rather than having fuzzy boundaries; I think we clipped these to whatever boundary was known rather than trying to handle the fuzzy nature), does handle things like introns effectively, does feature extraction effectively, but doesn’t do stuff that has no clear relevance like on-the-fly boolean calculations of presence/absence in feature sets on a per-nucleotide basis, that’s probably where the sweet spot is. But I’m interested in hearing counter-arguments.



> On Jan 25, 2016, at 1:32 PM, Evan Bolyen <notifications@github.com> wrote:
> 
> How did BioPython/PyCogent solve this problem? More importantly what kinds of operations did their models enable? Ideally we would want to avoid a world of for loops and property mining when querying sequence features.
> 
>